src/common/FileSystem.cpp

```diff
@@ -1253,6 +1268,143 @@ const std::vector<LoadedPakInfo>& GetLoadedPaks()
 	return loadedPaks;
 }
 
+// ExpandSymlink
+// code adapted from DarkPlaces fs.c FS_OpenReadFile
+// commit 1c55fd95ff34d9aafb76c927215ea8fbab5210c6
+// Copyright 2008 divverent
+// GPLv2+
+std::string ExpandSymlink(Str::StringRef path, Str::StringRef content)
+{
+		const char *filename = path.c_str();
+
+		char linkbuf[MAX_QPATH];
+		Q_strncpyz(linkbuf, content.c_str(), MAX_QPATH);
+		int count = strlen(linkbuf);
+
+		const char *mergeslash;
+		char *mergestart;
+
+		// Now combine the paths...
+		mergeslash = strrchr(filename, '/');
+		mergestart = linkbuf;
+		if(!mergeslash) {
+			mergeslash = filename;
+		}
+		while(!strncmp(mergestart, "../", 3)) {
+			mergestart += 3;
+			while(mergeslash > filename) {
+				--mergeslash;
+				if(*mergeslash == '/') {
+					break;
+				}
+			}
+		}
+		// Now, mergestart will point to the path to be appended, and mergeslash points to where it should be appended
+		if(mergeslash == filename) {
+			// Either mergeslash == filename, then we just replace the name (done below)
+		}
+		else {
+			// Or, we append the name after mergeslash;
+			// or rather, we can also shift the linkbuf so we can put everything up to and including mergeslash first
+			int spaceNeeded = mergeslash - filename + 1;
+			int spaceRemoved = mergestart - linkbuf;
+			if(count - spaceRemoved + spaceNeeded >= MAX_QPATH)
+			{
+				Log::Debug("symlink: too long path rejected\n");
+				return "";
+			}
+			memmove(linkbuf + spaceNeeded, linkbuf + spaceRemoved, count - spaceRemoved);
+			memcpy(linkbuf, filename, spaceNeeded);
+			linkbuf[count - spaceRemoved + spaceNeeded] = 0;
+			mergestart = linkbuf;
+		}
+
+		Log::Debug("found symlink: %s → %s", path, mergestart);
+
+		/*
+		if(CheckNastyPath (mergestart, false))
+		{
+			Log::Debug("symlink: nasty path %s rejected\n", mergestart);
+			return "";
+		}
+		*/
+
+		return mergestart;
+}
+
+std::string ReadZipFile(LoadedPakInfo pak, ZipArchive *zipFile, Str::StringRef originalPath, Str::StringRef path, int symlinkLevels, std::error_code& err)
+{
+	auto it = fileMap.find(path);
```

Comment:
`fileMap` is a global variable containing files for all paks so it doesn't make sense to be accessing it for resolving links within one zip archive.

Reply:
```c
// Map of filenames to pak files. The size_t is an offset into loadedPaks and
// the offset_t is the position within the zip archive (unused for PAK_DIR).
static std::unordered_map<std::string, std::pair<uint32_t, offset_t>> fileMap;
```

`fileMap.find(path)->second.second` is the offset to the content of the file with `path` in zip archive.

I haven't found a better way to load a file content without reparsing the zip again.

This is problematic if findMap returns a file that matches the path in another pak. I don't know if that can happen.

Reply:
`findMap` references a file that matches the path in another pak if the file is only available in the other pak (or take precedence with DEPS), so accessing `findMap` for that purpose is safe.

It maybe not very efficient, but it's OK for a compatibility layer. Doing this compute in zip internals is likely to add hard to read and maintain code. The need for a compatibility layer is probably not worth messing with zip internals.